### PR TITLE
Allow capture templates to insert at beginning or end of file

### PR DESCRIPTION
**Pull request by aspiers**
~~Currently based on #344 (eslint PR) so that I can use eslint to verify the changes are clean.~~ **Now rebased**

If no headers are in the header path list, insert at the end of the file, or the beginning if the Prepend option is set.

Lightly tested.

### CODE REVIEW

**Comment by aspiers** [comment]
Added tests and rebased on my `eslint` branch so that I could use `eslint`.

**Comment by aspiers** [comment]
@munen I can't figure out why [this undoable test is failing](https://app.circleci.com/pipelines/github/aspiers/organice/60/workflows/639f290d-0cfe-432a-a22b-bc13579c72c5/jobs/70/steps) - any ideas?

**Comment by munen** [comment]
> @munen I can't figure out why this undoable test is failing - any ideas?

Wow, you're asking about a failing test in a 42 changes PR that I haven't looked at, yet, because it's marked as a draft. No, at this point, I also don't know why the test is failing, but it looks to be the last commit(;

Yeah, I know, that's not helping at all. Sorry about that.

What about splitting the PR into two - one that you've manually verified and tested and that we can review. And another one adding more tests on top of and see what's going on? This particular PR seems to have a lot of changes that are not directly related to the title of it.

**Comment by aspiers** [comment]
[**@munen**](https://github.com/munen) commented on [June 9, 2020 7:39 PM](https://github.com/200ok-ch/organice/pull/324#issuecomment-641499027 "2020-06-09T18:39:40Z - Replied by Github Reply Comments"):
> > [@munen](https://github.com/munen) I can't figure out why this undoable test is failing - any ideas?

Already discussed on Matrix, but just for the record in case anyone else is reading here:

> Wow, you're asking about a failing test in a 42 changes PR that I haven't looked at, yet, because it's marked as a draft.

[As mentioned](https://github.com/200ok-ch/organice/pull/324#issuecomment-641485693), the vast majority of these changes are because it's based on (an old version of) my `eslint` branch.  Only the top few commits are relevant, and all but one of these are tiny.

> No, at this point, I also don't know why the test is failing, but it looks to be the last commit(;
> 
> Yeah, I know, that's not helping at all. Sorry about that.
> 
> What about splitting the PR into two - one that you've manually verified and tested and that we can review. And another one adding more tests on top of and see what's going on?

No need - the `eslint` branch was fully passing and unrelated to this.

> This particular PR seems to have a lot of changes that are not directly related to the title of it.

Yep, as explained [above](https://github.com/200ok-ch/organice/pull/324#issuecomment-641485693), I wanted to be able to use `eslint` when working on this branch.

Unfortunately after a lot more experiments, I still have zero clue why the final undo test is failing.  As agreed on Matrix, we'll just skip that test for now :-/

**Comment by aspiers** [comment]
Now rebased and ready for review.

File: src/components/OrgFile/components/CaptureModal/index.js [hunk 1]
```diff
@@ -22,10 +22,13 @@ export default ({ template, onCapture, headers, onClose }) => {
...
-    headers,
-    template,
-  ]);
+  const targetHeader = useMemo(
+    () =>
+      template.get('headerPaths').size === 0
+        ? 'FILE'
+        : headerWithPath(headers, template.get('headerPaths')),
+    [headers, template]
+  );
```
[the remark this answers is not in the document]
Comment (schoettl) [reply src/components/OrgFile/components/CaptureModal/index.js:31]: Oh, I got it ^^
`1, [2,3]` evaluates to `[2,3]`, but here the `[...]` is not part of the lambda expression. It's the second argument to `useMemo()`.

So, `targetHeader` can be `'FILE'`, a header object, or `null`.

[the remark this answers is not in the document]
Comment (aspiers) [reply src/components/OrgFile/components/CaptureModal/index.js:31]: Right :-)  And notice that this is just a reformatting by `prettier`, not a code change.

[the remark this answers is not in the document]
Comment (aspiers) [reply src/components/OrgFile/components/CaptureModal/index.js:31]: Sorry I'm getting confused and talking rubbish, clearly this *is* a code change!

File: src/reducers/org.js [hunk 2]
```diff
@@ -729,27 +729,51 @@ const insertCapture = (state, action) => {
...
   const { template, content, shouldPrepend } = action;
 
-  const parentHeader = headerWithPath(headers, template.get('headerPaths'));
-  if (!parentHeader) {
+  const { newIndex, nestingLevel, parentHeader } = insertCapturePosition(
+    template,
+    headers,
+    shouldPrepend
+  );
+  if (newIndex === null) {
```
Comment (aspiers) [inline src/reducers/org.js:737]: Oops, this should be `undefined`.

File: src/reducers/org.js [hunk 3]
```diff
@@ -729,27 +729,51 @@ const insertCapture = (state, action) => {
...
   );
 
-  const parentHeaderIndex = indexOfHeaderWithId(headers, parentHeader.get('id'));
-  const numSubheaders = numSubheadersOfHeaderWithId(headers, parentHeader.get('id'));
-  const newIndex = parentHeaderIndex + 1 + (shouldPrepend ? 0 : numSubheaders);
-
   state = state.update('headers', (headers) => headers.insert(newIndex, newHeader));
-
-  state = updateCookiesOfHeaderWithId(state, parentHeader.get('id'));
+  if (parentHeader !== undefined) {
```
[the remark this answers is not in the document]
Comment (aspiers) [reply src/reducers/org.js:747]: No, `newIndex === undefined` means that no target header could be found.  If `parentHeader !== undefined`, that means that we *do* have a target header to insert under, and if it `=== undefined`, that means we're inserting at the top / bottom of the file.

File: src/reducers/org.unit.test.js [hunk 4]
```diff
@@ -1,29 +1,42 @@
...
+      })
+      .toJS();
+  }
+
   beforeEach(() => {
-    state = fromJS(readInitialState());
-    state = state.setIn(['org', 'present'], parseOrg(testOrgFile));
-    store = createStore(undoable(reducer), state.getIn(['org', 'present']));
+    state = readInitialState();
+    state.org.present = parseOrg(testOrgFile);
```
[the remark this answers is not in the document]
Comment (aspiers) [reply src/reducers/org.unit.test.js:37]: > But now it's a bit inconsistent and differs from the "real world" state outside the tests, right?

Nope, in the live app, `state` and `state.org` are vanilla Javascript objects.  `state.org.present` is an Immutable `Map`.  So this change is not just necessary in order to allow testing of the new feature, but it's more consistent with the real world state.  It took me hours to figure that out :-(

[the remark this answers is not in the document]
Comment (aspiers) [reply src/reducers/org.unit.test.js:37]: > Nope, in the live app, `state` and `state.org` are vanilla Javascript objects.  `state.org.present` is an Immutable `Map`.

BTW you can see that clearly here:

https://github.com/200ok-ch/organice/blob/106f8cc27e9bd623689417e77b037e387041a19d/src/util/settings_persister.js#L188-L202

File: src/reducers/org.js (hunk 2 above)
Comment (aspiers) [reply src/reducers/org.js]: Fixed.

File: src/reducers/org.unit.test.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (schoettl) [reply src/reducers/org.unit.test.js:37]: Ah, alright, sorry.

File: src/reducers/org.unit.test.js [hunk 5]
```diff
@@ -96,11 +92,98 @@ describe('org reducer', () => {
...
+    it('should insert at bottom of file', () => {
+      const newHeaders = insertCapture(false);
+      expectOrigFirstHeader(newHeaders);
+      const last = newHeaders.last();
+      expect(last.getIn(['titleLine', 'rawTitle'])).toEqual('My task');
+      expect(last.getIn(['titleLine', 'todoKeyword'])).toEqual('TODO');
+      expect(last.get('rawDescription')).toEqual('Some description\n');
+    });
+
+    it.skip('is undoable', () => {
```
Comment (munen) [inline src/reducers/org.unit.test.js:174]: I don't understand why, yet, but the problem seems to be a racing condition or mutable state somewhere. When only executing this test, it works fine.

**Comment by munen** [comment]
Cool new feature, indeed!

I've added some docs to sample.org, documented the skipped test (and allowed it in eslint) and added attribution in the changelog.

Good work, @aspiers :pray:

**Pull request merged**